test/page.spec.js

```diff
@@ -781,6 +781,126 @@ module.exports.addTests = function({testRunner, expect, puppeteer, DeviceDescrip
     });
   });
 
+  describe('Page.waitForRequest', function() {
+    it('should work', async({page, server}) => {
+      await page.goto(server.EMPTY_PAGE);
+
+      server.setRoute('/getRequest', (req, res) => {
+        res.statusCode = 200;
+        res.end();
+      });
+
+      page.evaluate(() => {
```

Comment:
In order to avoid race condition between evaluation completion and page closing in the end of the test, we should `await` all our promises that interact with the page.

With the comment above, I'd write this test in the following way:

```js
it('should work', async({page, server}) => {
  await page.goto(server.EMPTY_PAGE);
  const [request] = await Promise.all([
    page.waitForRequest(server.PREFIX + '/404'),
    page.evaluate(() => fetch('/404', {method: 'GET'}))
  ]);
  expect(request.method()).toBe('GET');
  expect(request.url()).toBe(server.PREFIX + '/404');
});
```